## Preprocessing

In [1]:
# Import our dependencies
import sklearn as skl
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import StandardScaler
import pandas as pd
import tensorflow as tf

#  Import and read the charity_data.csv.
import pandas as pd
application_df = pd.read_csv("https://static.bc-edx.com/data/dl-1-2/m21/lms/starter/charity_data.csv")
application_df.head()

EIN                                      NAME APPLICATION_TYPE  \
0  10520599              BLUE KNIGHTS MOTORCYCLE CLUB              T10   
1  10531628    AMERICAN CHESAPEAKE CLUB CHARITABLE TR               T3   
2  10547893        ST CLOUD PROFESSIONAL FIREFIGHTERS               T5   
3  10553066            SOUTHSIDE ATHLETIC ASSOCIATION               T3   
4  10556103  GENETIC RESEARCH INSTITUTE OF THE DESERT               T3   

        AFFILIATION CLASSIFICATION      USE_CASE  ORGANIZATION  STATUS  \
0       Independent          C1000    ProductDev   Association       1   
1       Independent          C2000  Preservation  Co-operative       1   
2  CompanySponsored          C3000    ProductDev   Association       1   
3  CompanySponsored          C2000  Preservation         Trust       1   
4       Independent          C1000     Heathcare         Trust       1   

      INCOME_AMT SPECIAL_CONSIDERATIONS  ASK_AMT  IS_SUCCESSFUL  
0              0                      N     5000              1  
1         1-9999                      N   108590              1  
2              0                      N     5000              0  
3    10000-24999                      N     6692              1  
4  100000-499999                      N   142590              1

In [2]:
# Drop the non-beneficial ID columns, 'EIN' and 'NAME'.
application_df.drop(['EIN', 'NAME'], axis=1, inplace=True)
application_df

APPLICATION_TYPE       AFFILIATION CLASSIFICATION      USE_CASE  \
0                  T10       Independent          C1000    ProductDev   
1                   T3       Independent          C2000  Preservation   
2                   T5  CompanySponsored          C3000    ProductDev   
3                   T3  CompanySponsored          C2000  Preservation   
4                   T3       Independent          C1000     Heathcare   
...                ...               ...            ...           ...   
34294               T4       Independent          C1000    ProductDev   
34295               T4  CompanySponsored          C3000    ProductDev   
34296               T3  CompanySponsored          C2000  Preservation   
34297               T5       Independent          C3000    ProductDev   
34298               T3       Independent          C1000  Preservation   

       ORGANIZATION  STATUS     INCOME_AMT SPECIAL_CONSIDERATIONS   ASK_AMT  \
0       Association       1              0                      N      5000   
1      Co-operative       1         1-9999                      N    108590   
2       Association       1              0                      N      5000   
3             Trust       1    10000-24999                      N      6692   
4             Trust       1  100000-499999                      N    142590   
...             ...     ...            ...                    ...       ...   
34294   Association       1              0                      N      5000   
34295   Association       1              0                      N      5000   
34296   Association       1              0                      N      5000   
34297   Association       1              0                      N      5000   
34298  Co-operative       1          1M-5M                      N  36500179   

       IS_SUCCESSFUL  
0                  1  
1                  1  
2                  0  
3                  1  
4                  1  
...              ...  
34294              0  
34295              0  
34296              0  
34297              1  
34298              0  

[34299 rows x 10 columns]

In [3]:
# Determine the number of unique values in each column.
application_df.nunique()

APPLICATION_TYPE            17
AFFILIATION                  6
CLASSIFICATION              71
USE_CASE                     5
ORGANIZATION                 4
STATUS                       2
INCOME_AMT                   9
SPECIAL_CONSIDERATIONS       2
ASK_AMT                   8747
IS_SUCCESSFUL                2
dtype: int64

In [4]:
# Look at APPLICATION_TYPE value counts to identify and replace with "Other"
application_df['APPLICATION_TYPE'].value_counts()

APPLICATION_TYPE
T3     27037
T4      1542
T6      1216
T5      1173
T19     1065
T8       737
T7       725
T10      528
T9       156
T13       66
T12       27
T2        16
T25        3
T14        3
T29        2
T15        2
T17        1
Name: count, dtype: int64

In [5]:
# Choose a cutoff value and create a list of application types to be replaced
# use the variable name `application_types_to_replace`
application_types_to_replace = ['T9', 'T13', 'T12', 'T2', 'T25', 'T14', 'T29', 'T15', 'T17']

# Replace in dataframe
for app in application_types_to_replace:
    application_df['APPLICATION_TYPE'] = application_df['APPLICATION_TYPE'].replace(app,"Other")

# Check to make sure replacement was successful
application_df['APPLICATION_TYPE'].value_counts()

APPLICATION_TYPE
T3       27037
T4        1542
T6        1216
T5        1173
T19       1065
T8         737
T7         725
T10        528
Other      276
Name: count, dtype: int64

In [6]:
# Look at CLASSIFICATION value counts to identify and replace with "Other"
classification_counts = application_df['CLASSIFICATION'].value_counts()
classification_counts

CLASSIFICATION
C1000    17326
C2000     6074
C1200     4837
C3000     1918
C2100     1883
         ...  
C4120        1
C8210        1
C2561        1
C4500        1
C2150        1
Name: count, Length: 71, dtype: int64

In [7]:
# You may find it helpful to look at CLASSIFICATION value counts >1
filtered_classification_counts = classification_counts[classification_counts > 1]
filtered_classification_counts

CLASSIFICATION
C1000    17326
C2000     6074
C1200     4837
C3000     1918
C2100     1883
C7000      777
C1700      287
C4000      194
C5000      116
C1270      114
C2700      104
C2800       95
C7100       75
C1300       58
C1280       50
C1230       36
C1400       34
C7200       32
C2300       32
C1240       30
C8000       20
C7120       18
C1500       16
C1800       15
C6000       15
C1250       14
C8200       11
C1238       10
C1278       10
C1235        9
C1237        9
C7210        7
C2400        6
C1720        6
C4100        6
C1257        5
C1600        5
C1260        3
C2710        3
C0           3
C3200        2
C1234        2
C1246        2
C1267        2
C1256        2
Name: count, dtype: int64

In [8]:
# Choose a cutoff value and create a list of classifications to be replaced
# use the variable name `classifications_to_replace`
classifications_to_replace = ['C7000', 'C1700', 'C4000', 'C5000', 'C1270', 'C2700', 'C2800', 'C7100', 'C1300', 'C1280',
                              'C1230', 'C1400', 'C7200', 'C2300', 'C1240', 'C8000', 'C7120', 'C1500', 'C1800', 'C6000',
                              'C1250', 'C8200', 'C1238', 'C1278', 'C1235', 'C1237', 'C7210', 'C2400', 'C1720', 'C4100',
                              'C1257', 'C1600', 'C1260', 'C2710', 'C0', 'C3200', 'C1234', 'C1246', 'C1267', 'C1256',
                              'C1370', 'C1236', 'C2190', 'C4200', 'C5200', 'C1248', 'C2600', 'C1580', 'C6100', 'C1820',
                              'C1900', 'C2570', 'C1570', 'C3700', 'C2500', 'C1283', 'C2380', 'C1732', 'C1728', 'C2170',
                              'C4120', 'C8210', 'C4500', 'C1245', 'C2561', 'C2150']

# Replace in dataframe
for cls in classifications_to_replace:
    application_df['CLASSIFICATION'] = application_df['CLASSIFICATION'].replace(cls,"Other")

# Check to make sure replacement was successful
application_df['CLASSIFICATION'].value_counts()

CLASSIFICATION
C1000    17326
C2000     6074
C1200     4837
Other     2261
C3000     1918
C2100     1883
Name: count, dtype: int64

In [9]:
# Convert categorical data to numeric with `pd.get_dummies`
application_df = pd.get_dummies(application_df, columns=['APPLICATION_TYPE', 'AFFILIATION', 'CLASSIFICATION', 'INCOME_AMT', 'USE_CASE', 'ORGANIZATION', 'STATUS', 'SPECIAL_CONSIDERATIONS', 'IS_SUCCESSFUL'])
application_df

ASK_AMT  APPLICATION_TYPE_Other  APPLICATION_TYPE_T10  \
0          5000                   False                  True   
1        108590                   False                 False   
2          5000                   False                 False   
3          6692                   False                 False   
4        142590                   False                 False   
...         ...                     ...                   ...   
34294      5000                   False                 False   
34295      5000                   False                 False   
34296      5000                   False                 False   
34297      5000                   False                 False   
34298  36500179                   False                 False   

       APPLICATION_TYPE_T19  APPLICATION_TYPE_T3  APPLICATION_TYPE_T4  \
0                     False                False                False   
1                     False                 True                False   
2                     False                False                False   
3                     False                 True                False   
4                     False                 True                False   
...                     ...                  ...                  ...   
34294                 False                False                 True   
34295                 False                False                 True   
34296                 False                 True                False   
34297                 False                False                False   
34298                 False                 True                False   

       APPLICATION_TYPE_T5  APPLICATION_TYPE_T6  APPLICATION_TYPE_T7  \
0                    False                False                False   
1                    False                False                False   
2                     True                False                False   
3                    False                False                False   
4                    False                False                False   
...                    ...                  ...                  ...   
34294                False                False                False   
34295                False                False                False   
34296                False                False                False   
34297                 True                False                False   
34298                False                False                False   

       APPLICATION_TYPE_T8  ...  ORGANIZATION_Association  \
0                    False  ...                      True   
1                    False  ...                     False   
2                    False  ...                      True   
3                    False  ...                     False   
4                    False  ...                     False   
...                    ...  ...                       ...   
34294                False  ...                      True   
34295                False  ...                      True   
34296                False  ...                      True   
34297                False  ...                      True   
34298                False  ...                     False   

       ORGANIZATION_Co-operative  ORGANIZATION_Corporation  \
0                          False                     False   
1                           True                     False   
2                          False                     False   
3                          False                     False   
4                          False                     False   
...                          ...                       ...   
34294                      False                     False   
34295                      False                     False   
34296                      False                     False   
34297                      False                     False   
34298                       True                     False   

       

In [10]:
# Split our preprocessed data into our features and target arrays
y = application_df["IS_SUCCESSFUL_1"].values
X = application_df.drop(["IS_SUCCESSFUL_1", "IS_SUCCESSFUL_0"], axis=1).values

# Split the preprocessed data into a training and testing dataset
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=78)

In [11]:
# Create a StandardScaler instances
scaler = StandardScaler()

# Fit the StandardScaler
X_scaler = scaler.fit(X_train)

# Scale the data
X_train_scaled = X_scaler.transform(X_train)
X_test_scaled = X_scaler.transform(X_test)

## Optimize the Model

#### 1. Get the best model hyperparameters

In [12]:
!pip install keras-tuner

In [13]:
# Create a method that creates a new Sequential model with hyperparameter options
number_input_features = len(X_train[0])

def create_model(hp):
    nn_model = tf.keras.models.Sequential()

    # Allow kerastuner to decide which activation function to use in hidden layers
    activation = hp.Choice('activation',['relu','tanh','sigmoid'])

    # Allow kerastuner to decide number of neurons in first layer
    nn_model.add(tf.keras.layers.Dense(units=hp.Int('first_units',
        min_value=1,
        max_value=10,
        step=2), activation=activation, input_dim=number_input_features))

    # Allow kerastuner to decide number of hidden layers and neurons in hidden layers
    for i in range(hp.Int('num_layers', 1, 6)):
        nn_model.add(tf.keras.layers.Dense(units=hp.Int('units_' + str(i),
            min_value=1,
            max_value=10,
            step=2),
            activation=activation))

    nn_model.add(tf.keras.layers.Dense(units=1, activation="sigmoid"))

    # Compile the model
    nn_model.compile(loss="binary_crossentropy", optimizer='adam', metrics=["accuracy"])

    return nn_model

In [14]:
# Import the kerastuner library
import keras_tuner as kt

tuner = kt.Hyperband(
    create_model,
    objective="val_accuracy",
    max_epochs=20,
    hyperband_iterations=2)

Reloading Tuner from ./untitled_project/tuner0.json


In [15]:
# Run the kerastuner search for best hyperparameters
tuner.search(X_train_scaled,y_train,epochs=20,validation_data=(X_test_scaled,y_test))

In [16]:
# Evaluate best model against full test data
best_model = tuner.get_best_models(1)[0]
model_loss, model_accuracy = best_model.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

/Applications/anaconda3/envs/dev/lib/python3.12/site-packages/keras/src/layers/core/dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)
/Applications/anaconda3/envs/dev/lib/python3.12/site-packages/keras/src/saving/saving_lib.py:719: UserWarning: Skipping variable loading for optimizer 'adam', because it has 2 variables whereas the saved optimizer has 30 variables. 
  saveable.load_own_variables(weights_store.get(inner_path))


268/268 - 0s - 634us/step - accuracy: 0.7296 - loss: 0.5594
Loss: 0.5594413876533508, Accuracy: 0.7295626997947693


#### 2. Add hidden layers

In [17]:
# Define the model - deep neural net, i.e., the number of input features and hidden nodes for each layer.
number_input_features = len(X_train[0])
hidden_nodes_layer1 =  8
hidden_nodes_layer2 = 5
hidden_nodes_layer3 = 5
hidden_nodes_layer4 = 5
hidden_nodes_layer5 = 5

nn = tf.keras.models.Sequential()

# First hidden layer
nn.add(
    tf.keras.layers.Dense(units=hidden_nodes_layer1, input_dim=number_input_features, activation="relu")
)

# Second hidden layer
nn.add(tf.keras.layers.Dense(units=hidden_nodes_layer2, activation="relu"))

# Third hidden layer
nn.add(tf.keras.layers.Dense(units=hidden_nodes_layer3, activation="relu"))

# Fourth hidden layer
nn.add(tf.keras.layers.Dense(units=hidden_nodes_layer4, activation="relu"))

# Fifth hidden layer
nn.add(tf.keras.layers.Dense(units=hidden_nodes_layer5, activation="relu"))

# Output layer
nn.add(tf.keras.layers.Dense(units=1, activation="sigmoid"))

# Check the structure of the model
nn.summary()

Model: "sequential_1"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ dense_7 (Dense)                 │ (None, 8)              │           360 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_8 (Dense)                 │ (None, 5)              │            45 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_9 (Dense)                 │ (None, 5)              │            30 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_10 (Dense)                │ (None, 5)              │            30 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_11 (Dense)                │ (None, 5)              │            30 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_12 (Dense)                │ (None, 1)              │             6 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 501 (1.96 KB)

 Trainable params: 501 (1.96 KB)

 Non-trainable params: 0 (0.00 B)

In [18]:
# Compile the model
nn.compile(loss="binary_crossentropy", optimizer="adam", metrics=["accuracy"])

In [19]:
# Train the model
fit_model = nn.fit(X_train_scaled,y_train,epochs=5)

Epoch 1/5
804/804 ━━━━━━━━━━━━━━━━━━━━ 1s 282us/step - accuracy: 0.5730 - loss: 0.6645
Epoch 2/5
804/804 ━━━━━━━━━━━━━━━━━━━━ 0s 290us/step - accuracy: 0.7240 - loss: 0.5744
Epoch 3/5
804/804 ━━━━━━━━━━━━━━━━━━━━ 0s 282us/step - accuracy: 0.7335 - loss: 0.5544
Epoch 4/5
804/804 ━━━━━━━━━━━━━━━━━━━━ 0s 296us/step - accuracy: 0.7279 - loss: 0.5583
Epoch 5/5
804/804 ━━━━━━━━━━━━━━━━━━━━ 0s 274us/step - accuracy: 0.7273 - loss: 0.5587


In [20]:
# Evaluate the model using the test data
model_loss, model_accuracy = nn.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

268/268 - 0s - 347us/step - accuracy: 0.7227 - loss: 0.5573
Loss: 0.5573304891586304, Accuracy: 0.7226822376251221


#### 3. Add more neurons to hidden layers

In [21]:
# Define the model - deep neural net, i.e., the number of input features and hidden nodes for each layer.
number_input_features = len(X_train[0])
hidden_nodes_layer1 =  15
hidden_nodes_layer2 = 15
hidden_nodes_layer3 = 15
hidden_nodes_layer4 = 15
hidden_nodes_layer5 = 15

nn = tf.keras.models.Sequential()

# First hidden layer
nn.add(
    tf.keras.layers.Dense(units=hidden_nodes_layer1, input_dim=number_input_features, activation="relu")
)

# Second hidden layer
nn.add(tf.keras.layers.Dense(units=hidden_nodes_layer2, activation="relu"))

# Third hidden layer
nn.add(tf.keras.layers.Dense(units=hidden_nodes_layer3, activation="relu"))

# Fourth hidden layer
nn.add(tf.keras.layers.Dense(units=hidden_nodes_layer4, activation="relu"))

# Fifth hidden layer
nn.add(tf.keras.layers.Dense(units=hidden_nodes_layer5, activation="relu"))

# Output layer
nn.add(tf.keras.layers.Dense(units=1, activation="sigmoid"))

# Check the structure of the model
nn.summary()

Model: "sequential_2"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ dense_13 (Dense)                │ (None, 15)             │           675 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_14 (Dense)                │ (None, 15)             │           240 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_15 (Dense)                │ (None, 15)             │           240 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_16 (Dense)                │ (None, 15)             │           240 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_17 (Dense)                │ (None, 15)             │           240 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_18 (Dense)                │ (None, 1)              │            16 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 1,651 (6.45 KB)

 Trainable params: 1,651 (6.45 KB)

 Non-trainable params: 0 (0.00 B)

In [22]:
# Compile the model
nn.compile(loss="binary_crossentropy", optimizer="adam", metrics=["accuracy"])

In [23]:
# Train the model
fit_model = nn.fit(X_train_scaled,y_train,epochs=5)

Epoch 1/5
804/804 ━━━━━━━━━━━━━━━━━━━━ 1s 294us/step - accuracy: 0.6825 - loss: 0.6115
Epoch 2/5
804/804 ━━━━━━━━━━━━━━━━━━━━ 0s 303us/step - accuracy: 0.7309 - loss: 0.5552
Epoch 3/5
804/804 ━━━━━━━━━━━━━━━━━━━━ 0s 283us/step - accuracy: 0.7286 - loss: 0.5551
Epoch 4/5
804/804 ━━━━━━━━━━━━━━━━━━━━ 0s 282us/step - accuracy: 0.7365 - loss: 0.5468
Epoch 5/5
804/804 ━━━━━━━━━━━━━━━━━━━━ 0s 301us/step - accuracy: 0.7307 - loss: 0.5514


In [24]:
# Evaluate the model using the test data
model_loss, model_accuracy = nn.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

268/268 - 0s - 345us/step - accuracy: 0.7245 - loss: 0.5593
Loss: 0.55934739112854, Accuracy: 0.7245481014251709


#### 4. Add the number of epochs

In [25]:
# Define the model - deep neural net, i.e., the number of input features and hidden nodes for each layer.
number_input_features = len(X_train[0])
hidden_nodes_layer1 =  15
hidden_nodes_layer2 = 15
hidden_nodes_layer3 = 15
hidden_nodes_layer4 = 15
hidden_nodes_layer5 = 15

nn = tf.keras.models.Sequential()

# First hidden layer
nn.add(
    tf.keras.layers.Dense(units=hidden_nodes_layer1, input_dim=number_input_features, activation="relu")
)

# Second hidden layer
nn.add(tf.keras.layers.Dense(units=hidden_nodes_layer2, activation="relu"))

# Third hidden layer
nn.add(tf.keras.layers.Dense(units=hidden_nodes_layer3, activation="relu"))

# Fourth hidden layer
nn.add(tf.keras.layers.Dense(units=hidden_nodes_layer4, activation="relu"))

# Fifth hidden layer
nn.add(tf.keras.layers.Dense(units=hidden_nodes_layer5, activation="relu"))

# Output layer
nn.add(tf.keras.layers.Dense(units=1, activation="sigmoid"))

# Check the structure of the model
nn.summary()

Model: "sequential_3"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ dense_19 (Dense)                │ (None, 15)             │           675 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_20 (Dense)                │ (None, 15)             │           240 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_21 (Dense)                │ (None, 15)             │           240 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_22 (Dense)                │ (None, 15)             │           240 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_23 (Dense)                │ (None, 15)             │           240 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_24 (Dense)                │ (None, 1)              │            16 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 1,651 (6.45 KB)

 Trainable params: 1,651 (6.45 KB)

 Non-trainable params: 0 (0.00 B)

In [26]:
# Compile the model
nn.compile(loss="binary_crossentropy", optimizer="adam", metrics=["accuracy"])

In [27]:
# Train the model
fit_model = nn.fit(X_train_scaled,y_train,epochs=100)

Epoch 1/100
804/804 ━━━━━━━━━━━━━━━━━━━━ 1s 288us/step - accuracy: 0.6486 - loss: 0.6347
Epoch 2/100
804/804 ━━━━━━━━━━━━━━━━━━━━ 0s 282us/step - accuracy: 0.7326 - loss: 0.5546
Epoch 3/100
804/804 ━━━━━━━━━━━━━━━━━━━━ 0s 292us/step - accuracy: 0.7333 - loss: 0.5549
Epoch 4/100
804/804 ━━━━━━━━━━━━━━━━━━━━ 0s 282us/step - accuracy: 0.7329 - loss: 0.5497
Epoch 5/100
804/804 ━━━━━━━━━━━━━━━━━━━━ 0s 278us/step - accuracy: 0.7296 - loss: 0.5541
Epoch 6/100
804/804 ━━━━━━━━━━━━━━━━━━━━ 0s 276us/step - accuracy: 0.7309 - loss: 0.5506
Epoch 7/100
804/804 ━━━━━━━━━━━━━━━━━━━━ 0s 302us/step - accuracy: 0.7312 - loss: 0.5514
Epoch 8/100
804/804 ━━━━━━━━━━━━━━━━━━━━ 0s 279us/step - accuracy: 0.7383 - loss: 0.5438
Epoch 9/100
804/804 ━━━━━━━━━━━━━━━━━━━━ 0s 279us/step - accuracy: 0.7309 - loss: 0.5511
Epoch 10/100
804/804 ━━━━━━━━━━━━━━━━━━━━ 0s 280us/step - accuracy: 0.7322 - loss: 0.5451
Epoch 11/100
804/804 ━━━━━━━━━━━━━━━━━━━━ 0s 287us/step - accuracy: 0.7311 - loss: 0.5477
Epoch 12/100
804/80

In [28]:
# Evaluate the model using the test data
model_loss, model_accuracy = nn.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

268/268 - 0s - 355us/step - accuracy: 0.7248 - loss: 0.5542
Loss: 0.5542405843734741, Accuracy: 0.724781334400177


#### 5. Use different activation functions for the hidden layers

In [29]:
# Define the model - deep neural net, i.e., the number of input features and hidden nodes for each layer.
number_input_features = len(X_train[0])
hidden_nodes_layer1 =  15
hidden_nodes_layer2 = 15
hidden_nodes_layer3 = 15
hidden_nodes_layer4 = 15
hidden_nodes_layer5 = 15

nn = tf.keras.models.Sequential()

# First hidden layer
nn.add(
    tf.keras.layers.Dense(units=hidden_nodes_layer1, input_dim=number_input_features, activation="tanh")
)

# Second hidden layer
nn.add(tf.keras.layers.Dense(units=hidden_nodes_layer2, activation="tanh"))

# Third hidden layer
nn.add(tf.keras.layers.Dense(units=hidden_nodes_layer3, activation="tanh"))

# Fourth hidden layer
nn.add(tf.keras.layers.Dense(units=hidden_nodes_layer4, activation="tanh"))

# Fifth hidden layer
nn.add(tf.keras.layers.Dense(units=hidden_nodes_layer5, activation="tanh"))

# Output layer
nn.add(tf.keras.layers.Dense(units=1, activation="sigmoid"))

# Check the structure of the model
nn.summary()

Model: "sequential_4"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ dense_25 (Dense)                │ (None, 15)             │           675 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_26 (Dense)                │ (None, 15)             │           240 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_27 (Dense)                │ (None, 15)             │           240 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_28 (Dense)                │ (None, 15)             │           240 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_29 (Dense)                │ (None, 15)             │           240 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_30 (Dense)                │ (None, 1)              │            16 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 1,651 (6.45 KB)

 Trainable params: 1,651 (6.45 KB)

 Non-trainable params: 0 (0.00 B)

In [30]:
# Compile the model
nn.compile(loss="binary_crossentropy", optimizer="adam", metrics=["accuracy"])

In [31]:
# Train the model
fit_model = nn.fit(X_train_scaled,y_train,epochs=100)

Epoch 1/100
804/804 ━━━━━━━━━━━━━━━━━━━━ 1s 285us/step - accuracy: 0.6862 - loss: 0.6022
Epoch 2/100
804/804 ━━━━━━━━━━━━━━━━━━━━ 0s 277us/step - accuracy: 0.7263 - loss: 0.5623
Epoch 3/100
804/804 ━━━━━━━━━━━━━━━━━━━━ 0s 278us/step - accuracy: 0.7321 - loss: 0.5536
Epoch 4/100
804/804 ━━━━━━━━━━━━━━━━━━━━ 0s 294us/step - accuracy: 0.7296 - loss: 0.5542
Epoch 5/100
804/804 ━━━━━━━━━━━━━━━━━━━━ 0s 280us/step - accuracy: 0.7334 - loss: 0.5528
Epoch 6/100
804/804 ━━━━━━━━━━━━━━━━━━━━ 0s 277us/step - accuracy: 0.7290 - loss: 0.5571
Epoch 7/100
804/804 ━━━━━━━━━━━━━━━━━━━━ 0s 276us/step - accuracy: 0.7303 - loss: 0.5506
Epoch 8/100
804/804 ━━━━━━━━━━━━━━━━━━━━ 0s 294us/step - accuracy: 0.7311 - loss: 0.5498
Epoch 9/100
804/804 ━━━━━━━━━━━━━━━━━━━━ 0s 276us/step - accuracy: 0.7331 - loss: 0.5481
Epoch 10/100
804/804 ━━━━━━━━━━━━━━━━━━━━ 0s 278us/step - accuracy: 0.7331 - loss: 0.5447
Epoch 11/100
804/804 ━━━━━━━━━━━━━━━━━━━━ 0s 278us/step - accuracy: 0.7341 - loss: 0.5443
Epoch 12/100
804/80

In [32]:
# Evaluate the model using the test data
model_loss, model_accuracy = nn.evaluate(X_test_scaled,y_test,verbose=2)
print(f"Loss: {model_loss}, Accuracy: {model_accuracy}")

268/268 - 0s - 355us/step - accuracy: 0.7252 - loss: 0.5527
Loss: 0.5527417659759521, Accuracy: 0.7252478003501892


## Report on the Neural Network Model

##### 1. Overview
The analysis aimed at assisting the nonprofit foundation Alphabet Soup in selecting the applicants for funding with the best chance of success in their ventures.

#### 2. Results
##### 2.1 Data Preprocessing
The target variable was "IS_SUCCESSFUL_1" (the money was used effectively).

The feature variables were as follows:

- APPLICATION_TYPE — Alphabet Soup application type
- AFFILIATION — Affiliated sector of industry
- CLASSIFICATION — Government organization classification
- USE_CASE — Use case for funding
- ORGANIZATION — Organization type
- STATUS — Active status
- INCOME_AMT — Income classification
- SPECIAL_CONSIDERATIONS — Special considerations for application
- ASK_AMT — Funding amount requested

"IS_SUCCESSFUL_0" (the money was not used effectively) should be removed from the input data because they are neither targets nor features.

##### 2.2 Compiling, Training, and Evaluating the Model
The initial neural network model used the following features:

- Layers: 2
- Neurons: 8, 5
- Activation function: ReLU for input layer, sigmoid for output layer

ReLU was used as it is a fast, efficient, and simple activation function, particularly in deep networks. It works well in most hidden layers. Sigmoid was used as it is best suited for binary classification tasks to model probabilities, or when the output needs to be between 0 and 1.

The initial model had 72.3% accuracy. Despite the attempts to increase the model performance, no model could achieve the target 75% accuracy. The attempts and the respective performance are summarized as follows:

- Attempt 1 - Getting the best model hyperparameters: 73.0% accuracy
- Attempt 2 - Adding hidden layers from 3 to 5: 72.3% accuracy
- Attempt 3 - Adding neurons from a range of 5 and 8 to all 15: 72.5% accuracy
- Attempt 4 - Adding the number of epochs from 5 to 100: 72.8% accuracy
- Attempt 5 - Using tanh instead of ReLU as the activation function for the hidden layers: 72.5% accuracy

#### 3. Summary
The deep learning model was fairly effective in predicting applicants with the best chance of success in their ventures. The 72.3% accuracy means that for every 100 test instances, the model correctly predicted the outcome for 72.3 of them.

The model could be improved by modifying the feature variables if more context is given on the meaning of each variable. Minimizing the variables that are neither targets nor features could make the model more focused with less confusion.

In [33]:
# Export our model to HDF5 file
nn.save('AlphabetSoupCharity_Optimization_2.h5')